velox/functions/prestosql/window/AggregateWindow.cpp

```diff
@@ -0,0 +1,286 @@
+// for each row.
+class AggregateWindowFunction : public exec::WindowFunction {
+ public:
+  AggregateWindowFunction(
+      const std::string& name,
+      const std::vector<exec::WindowFunctionArg>& args,
+      const TypePtr& resultType,
+      velox::memory::MemoryPool* pool)
+      : WindowFunction(resultType, pool) {
+    argTypes_.reserve(args.size());
+    argIndices_.reserve(args.size());
+    argVectors_.reserve(args.size());
+    for (const auto& arg : args) {
+      argTypes_.push_back(arg.type);
+      // TODO : Enhance code to handle constant arguments.
+      VELOX_CHECK_NULL(
+          arg.constantValue,
+          "Constant arguments are not supported in aggregate window functions.");
+      VELOX_CHECK(arg.index.has_value());
+      argIndices_.push_back(arg.index.value());
+      argVectors_.push_back(BaseVector::create(arg.type, 0, pool_));
+    }
+    // Create an Aggregate function object to do result computation. Window
+    // function usage only requires single group aggregation for calculating
+    // the function value for each row.
+    aggregate_ = exec::Aggregate::create(
+        name, core::AggregationNode::Step::kSingle, argTypes_, resultType);
+
+    // Aggregate initialization.
+    // Row layout is:
+    //  - null flags - one bit per aggregate.
+    //  - uint32_t row size,
+    //  - fixed-width accumulators - one per aggregate
+    //
+    // Here we always make space for a row size since we only have one
+    // row and no RowContainer. We also have a single aggregate here, so there
+    // is only one null bit.
+    singleGroupRowSize_ = kRowSizeOffset + sizeof(int32_t);
+    aggregate_->setOffsets(
+        singleGroupRowSize_,
+        exec::RowContainer::nullByte(kNullOffset),
+        exec::RowContainer::nullMask(kNullOffset),
+        /* needed for out of line allocations */ kRowSizeOffset);
+    singleGroupRowSize_ += aggregate_->accumulatorFixedWidthSize();
+
+    // Constructing the single row in the MemoryPool.
+    singleGroupRow_ = (char*)(pool_->allocate(singleGroupRowSize_));
+
+    // Constructing a vector of a single result value used for copying from
+    // the aggregate to the final result.
+    aggregateResultVector_ = BaseVector::create(resultType, 1, pool_);
+  }
+
+  ~AggregateWindowFunction() {
+    std::vector<char*> singleGroupRowVector = {singleGroupRow_};
+    aggregate_->destroy(folly::Range(singleGroupRowVector.data(), 1));
+    pool_->free(singleGroupRow_, singleGroupRowSize_);
+  }
+
+  void resetPartition(const exec::WindowPartition* partition) override {
+    partition_ = partition;
+  }
+
+  void apply(
+      const BufferPtr& /*peerGroupStarts*/,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds,
+      vector_size_t resultOffset,
+      const VectorPtr& result) override {
+    int numRows = frameStarts->size() / sizeof(vector_size_t);
+    auto frameStartsVector = frameStarts->as<vector_size_t>();
```

Comment:
naming: rawFrameStarts and rawFrameEnds; these variables are raw pointers not std::vectors